presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataUpdater.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.facebook.airlift.concurrent.MoreFutures;
+import com.facebook.presto.spi.ConnectorMetadataUpdateHandle;
+import com.facebook.presto.spi.connector.ConnectorMetadataUpdater;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.SettableFuture;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+public class HiveMetadataUpdater
+        implements ConnectorMetadataUpdater
+{
+    private final Map<String, SettableFuture<String>> requestFutureMap = new ConcurrentHashMap<>();
+    private final Map<Integer, String> writerRequestMap = new ConcurrentHashMap<>();
+    private final Queue<HiveMetadataUpdateHandle> hiveMetadataRequestQueue = new ConcurrentLinkedQueue<>();
+
+    public HiveMetadataUpdater() {}
+
+    @Override
+    public List<ConnectorMetadataUpdateHandle> getPendingMetadataUpdateRequests()
+    {
+        ImmutableList.Builder<ConnectorMetadataUpdateHandle> result = new ImmutableList.Builder<>();
+        for (HiveMetadataUpdateHandle request : hiveMetadataRequestQueue) {
+            result.add(request);
+        }
+        return result.build();
+    }
+
+    @Override
+    public void setMetadataUpdateResults(List<ConnectorMetadataUpdateHandle> results)
```

Comment:
This looks like an expensive call directly executed from HTTP thread (from `TaskResource.updateMetadataResults`). Potentially, this will exhaust jetty threads. A good way is to make it async and move it to a CPU thread.